android/jni/com/mapswithme/platform/SocketImpl.cpp

```diff
@@ -0,0 +1,78 @@
+#include "../core/jni_helper.hpp"
+#include "base/logging.hpp"
+#include "platform/socket.hpp"
+
+namespace platform
+{
+class SocketImpl : public Socket
+{
+public:
+  SocketImpl()
+  {
+    JNIEnv * env = jni::GetEnv();
+    m_socketClass = jni::GetGlobalClassRef(env, "com/mapswithme/maps/location/SocketWrapper");
+    static jmethodID const socketConstructor = jni::GetConstructorID(env, m_socketClass, "()V");
+    jni::TScopedLocalRef localSelf(env, env->NewObject(m_socketClass, socketConstructor));
+    m_self = env->NewGlobalRef(localSelf.get());
+    ASSERT(m_self, ());
+  }
+
+  ~SocketImpl()
+  {
+    Close();
+    JNIEnv * env = jni::GetEnv();
+    env->DeleteGlobalRef(m_self);
+    env->DeleteGlobalRef(m_socketClass);
+  }
+
+  bool Open(string const & host, uint16_t port)
+  {
+    JNIEnv * env = jni::GetEnv();
+    static jmethodID const openMethod =
+        jni::GetMethodID(env, m_self, "open", "(Ljava/lang/String;I)Z");
+    return env->CallBooleanMethod(m_self, openMethod, jni::ToJavaString(env, host),
+                                  static_cast<jint>(port));
+  }
+
+  void Close()
+  {
+    JNIEnv * env = jni::GetEnv();
+    static jmethodID const closeMethod = jni::GetMethodID(env, m_self, "close", "()V");
+    env->CallVoidMethod(m_self, closeMethod);
+  }
+
+  bool Read(uint8_t * data, uint32_t count)
+  {
+    JNIEnv * env = jni::GetEnv();
+    jbyteArray array = env->NewByteArray(count);
+    static jmethodID const readMethod = jni::GetMethodID(env, m_self, "read", "([BI)B)");
+    jboolean result = env->CallBooleanMethod(m_self, readMethod, result, static_cast<jint>(count));
+    //this call copies java byte array to native buffer
+    env->GetByteArrayRegion(array, 0, count, (jbyte*)data);
```

Comment:
Fixed. The same should in 'Write' method? or if 'const' is used then 'static_cast' is no needed?